### PR TITLE
C++: IR data flow through global variables

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

**Comment by rdmarsh2** [comment]
Opened as draft since it builds on #8515

**Comment by rdmarsh2** [comment]
@geoffw0 there's some new false positives in the CWE-611 tests, and I'm not sure whether they're an issue with this PR or just exposing a known issue with the CWE-611 query. Could you take a look at those?

File: cpp/ql/test/query-tests/Security/CWE/CWE-497/semmle/tests/tests.cpp
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-497/semmle/tests/tests.cpp:73]: Excellent!

File: cpp/ql/test/query-tests/Security/CWE/CWE-497/semmle/tests/tests2.cpp
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-497/semmle/tests/tests2.cpp:82]: Great. :+1:

File: cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp
Comment (geoffw0) [inline cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp:45]: It looks like barriers aren't working for global variables?  I'm not 100% sure whether the problem lies in IR data flow or in the way this query defines its barriers.

I don't think we should tolerate these false positives (because they are likely to be more common than new good results in the wild), but if you can't easily fix them in the IR dataflow I will have a look at the query, and possibly just block results involving global variables for now.  That will undo all of the changes for this query, but keep the (valuable) improvements in the other queries.

Let me know if you think there's an IR dataflow fix first.

**Review by geoffw0 — COMMENTED** [review]
Reviewed the changes in CPP security query test results only.

File: cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp:45]: I _think_ the issue might be in the way we define transformers using GVN in `DomConfigurationSetParameterTransformer`. It's a totally reasonable way to do it, but since we don't try to do any alias analysis for global variables, the GVNs we get for uses of global variables will pretty much always be unique, which makes GVN for these values quite useless (please correct me if I'm misremembering what the IR actually looks like for uses of global variables @rdmarsh2).

I don't think we have any good way of doing GVN for global variables, so
I think the best solution is to disable detection in the case of global variables in the problematic XML libraries (I guess it's really all of them?).

If we had use-use dataflow (😭) we could use that instead of GVN, and that would support global variables as well.

Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp:45]: I've created https://github.com/rdmarsh2/ql/pull/5/files as a workaround, but I'm open to suggestions if we can do better.

Comment (rdmarsh2) [reply cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp:45]: The way this is implemented now, every read of a global variable has flow from every write to that variable (via the variable node). I think that's what's causing the issue here - the initial object creation *also* flows to the use. I'm not sure I can fix that with a reasonable amount of work, though...

**Comment by MathiasVP** [comment]
The performance on `vim/vim` looks like something we should investigate before we merge this.

**Review by jketema — COMMENTED** [review]
Two naive questions to start, just to further my understanding.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:103]: Because we're no longer talking (exclusively) about functions - we can now have code in global variable initializers that shows up in the dataflow graph, and there's places that assume `getFunction` has a result for those nodes.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:103]: Thanks for the explanation. I was aware of the related changes in https://github.com/github/codeql/pull/8515. It does make me wonder whether we should spend some time renaming predicates to better reflect what they're doing (doesn't have to be in this PR).

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:103]: I also see that in some places `func` was renamed to `decl` and in other places it wasn't. Is that on purpose?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:40]: I prefer to have this spelled out, as that's what's done for other predicates.
```suggestion
  predicate varHasIRFunction(GlobalOrNamespaceVariable var) {
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/IRConfiguration.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/IRConfiguration.qll:28]:
```suggestion
  predicate shouldEvaluateDebugStringsForFunction(Language::Declaration decl) { any() }
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/internal/TIRVariable.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/internal/TIRVariable.qll:5]:
```suggestion
  TIRUserVariable(Language::Variable var, Language::LanguageType type, Language::Declaration decl) {
```
Also in a few cases below. I'll stop commenting on this now. I just think this is more consistent. I don't have any problems off-loading this in a separate PR with is just some consistency fixes.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:43]: This, I assume, is to avoid some of the pref problems you were running into? Maybe worth a comment?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:45]: I assume these are constant too? Any reason why we do want to keep these in?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll:122]:
```suggestion
  not Raw::varHasIRFunction(expr.getEnclosingVariable())
```

**Review by jketema — COMMENTED** [review]
Next batch of comments. I'm slowly working my way through this. Feel free to go ahead and fixing things even though I'm not yet done (whatever your preferred way of working is).

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll:793]: That only applies to the translation of a `this` expression, which doesn't exist for global/namespace variables.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:43]: Mostly that constant integers will end up getting inlined in IR generation, so we don't need IR for them.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:45]: > Maybe that applies to all non-numeric types, though...

If that's true. What's the consequence of that at the moment? Just inconsistent IR that will be captured by the consistency check you also introduce in this PR? If so, I think this can be addressed later when we see that this is a problem.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:124]: Can these two cases actually happen?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:68]: Please fix.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:45]: If there were an interesting source in the initializer of a global variable with a struct or pointer type (probably flowing via a field), we wouldn't currently get taint flow from it.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:28]: Only the order matters for these, not the specific number. `0` would work just as well.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:45]: Ok, but that's equal to the situation we had before this PR, right?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:124]: Yes - there's a [GCC extension](https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html) that allows statements in an expression context.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:28]: Looking at the rest of the code, we generally seem to start at 0, but this is truly nitpicking, so feel free to ignore this.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedGlobalVar.qll:124]: Ok.

**Comment by jketema** [comment]
Looking at the DCA experiment: did you look whether the alert changes made sense?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll
Comment (rdmarsh2) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll:45]: Right.

**Comment by jketema** [comment]
@rdmarsh2 I think there's still the following to resolve:
* code scanning alerts
* open question regarding DCA
* merge conflicts

**Comment by jketema** [comment]
There's also still: https://github.com/github/codeql/pull/8596#discussion_r920049933 Is this something that needs to be looked at on the extractor side?

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected
Comment (rdmarsh2) [reply cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected]: This probably is something we need to look at eventually. The consistency check is new, and I think these don't have to do with the rest of the PR.

**Comment by rdmarsh2** [comment]
The DCA alert changes made sense, yes. Responded to that last comment - I'm not sure if they need work on the extractor side or if they need special handling in IR generation.

File: cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected
Comment (jketema) [reply cpp/ql/test/library-tests/syntax-zoo/aliased_ssa_consistency.expected]: Agreed that this doesn't have anything to do with this PR.

**Comment by jketema** [comment]
LGTM. I've kicked off a final DCA experiment, as we changed quite a few things related to (global) variables on the extractor side since this PR was opened. I'll approve once the experiment is done (unless we find some serious problem that is new compared to the previous DCA experiment, which probably isn't very likely).

**Comment by jketema** [comment]
New DCA results are in:
* There's a significant increase in the number of alerts, which makes sense given the extractor related issues we fixed and as we added/enabled systemd and Samate. The alters look ok, but a second pair of eyes would be welcome. 
* The second vim run OOM'ed this might either be related to the OOM problems we were having anyway, or due to a bad join order. Have the join orders been looked at?

**Comment by MathiasVP** [comment]
> There's a significant increase in the number of alerts (compared to the earlier DCA experiment), which makes sense given the extractor related issues we fixed and as we added/enabled systemd and Samate. The alerts look ok to me, but a second pair of eyes would be welcome.

So far:
- All the new results on `cpp/command-line-injection` are new TPs involving global variables 🎉

... I'll update this post as I look through the other new results.

**Comment by rdmarsh2** [comment]
Looks like the OOM is on ExecTainted.ql - all other queries finish before the error. I'll take a deeper look at it.

**Comment by jketema** [comment]
> Looks like the OOM is on ExecTainted.ql - all other queries finish before the error. I'll take a deeper look at it.

You might want to have a brief sync with Mathias about it. If you can't track it down I would suggest repeating the vim part of the DCA experiment with a **lower** `--ram` option than the `14979` currently used. I can explain via our internal channels.

**Comment by jketema** [comment]
I if see correctly, the re-run of DCA on vim was without the `--ram` option, because my directions were slightly wrong. Soory! Nevertheless, the re-run was successful, and the slowdown doesn't seem to have changed compared to the previous run. I propose we go ahead and merge this, and that if we see DCA OOMs again on vim, we move vim to a slightly bigger runner.

**Review by jketema — APPROVED** [review]

**Pull request merged**
